Windows golang build bug

```bash
In file included from .\llama_embedder.go:10:
./wrapper.h:6:10: fatal error: dlfcn.h: No such file or directory
    6 | #include <dlfcn.h>
      |          ^~~~~~~~~
compilation terminated.
```

--- a/.github/workflows/go-test.yml
+++ b/.github/workflows/go-test.yml
@@ -73,9 +73,10 @@ jobs:
           echo "CMAKE_BUILD_FLAGS=-j$(nproc)" >> $GITHUB_ENV
       - name: Set Windows Env
         if: matrix.os == 'windows-latest'
+        shell: pwsh
         run: |
-          echo "SHARED_LIB_PATH=../../build/Release" >> $GITHUB_ENV
-          echo "CMAKE_BUILD_FLAGS=-j${env:NUMBER_OF_PROCESSORS}" >> $GITHUB_ENV
+          Add-Content -Path $env:GITHUB_ENV -Value "SHARED_LIB_PATH=../../build/Release"
+          Add-Content -Path $env:GITHUB_ENV -Value "CMAKE_BUILD_FLAGS=-j$env:NUMBER_OF_PROCESSORS"
       - name: Build and Test
         run: |
           make go-test

--- a/.github/workflows/go-test.yml
+++ b/.github/workflows/go-test.yml
@@ -30,8 +30,9 @@ jobs:
           working-directory: bindings/go
           args: --verbose
   build:
-    runs-on: ${{matrix.os}}
-    needs: [lint]
+    name: Build and Test
+    runs-on: ${{ matrix.os }}
+    needs: [ lint ]
     strategy:
       matrix:
         os: [ ubuntu-latest, windows-latest, macos-latest ] #, windows-latest, macos-latest, macos-14, ubuntu-latest,
@@ -64,14 +65,17 @@ jobs:
         run: |
           echo "CMAKE_FLAGS=-DGGML_METAL_EMBED_LIBRARY=ON" >> $GITHUB_ENV
           echo "SHARED_LIB_PATH=../../build" >> $GITHUB_ENV
+          echo "CMAKE_BUILD_FLAGS=-j$(sysctl -n hw.logicalcpu)" >> $GITHUB_ENV
       - name: Set Linux Env
         if: matrix.os == 'ubuntu-latest'
         run: |
           echo "SHARED_LIB_PATH=../../build" >> $GITHUB_ENV
+          echo "CMAKE_BUILD_FLAGS=-j$(nproc)" >> $GITHUB_ENV
       - name: Set Windows Env
         if: matrix.os == 'windows-latest'
         run: |
           echo "SHARED_LIB_PATH=../../build/Release" >> $GITHUB_ENV
+          echo "CMAKE_BUILD_FLAGS=-j${env:NUMBER_OF_PROCESSORS}" >> $GITHUB_ENV
       - name: Build and Test
         run: |
           make go-test
@@ -79,3 +83,5 @@ jobs:
           GTEST_COLOR: yes
           GTEST_OUTPUT: xml:test_results.xml
           SHARED_LIB_PATH: ${{ env.SHARED_LIB_PATH }}
+          CMAKE_FLAGS: ${{ env.CMAKE_FLAGS }}
+          CMAKE_BUILD_FLAGS: ${{ env.CMAKE_BUILD_FLAGS }}

--- a/bindings/go/llama_embedder.go
+++ b/bindings/go/llama_embedder.go
@@ -5,8 +5,7 @@ package llama_embedder
 #cgo CXXFLAGS: -I. -std=c++11 -Wall -Wextra -pedantic
 
 #cgo darwin LDFLAGS: -stdlib=libc++
-#cgo linux !clang LDFLAGS: -ldl -lstdc++
-#cgo linux clang LDFLAGS: -ldl -stdlib=libc++
+#cgo linux LDFLAGS: -ldl -lstdc++
 #cgo windows LDFLAGS: -lkernel32
 #include <stdlib.h>
 #include "wrapper.h"

--- a/bindings/go/llama_embedder.go
+++ b/bindings/go/llama_embedder.go
@@ -305,7 +305,7 @@ func getOSSharedLibName() string {
 	case "darwin":
 		return "libllama-embedder.dylib"
 	case "windows":
-		return "libllama-embedder.dll"
+		return "llama-embedder.dll"
 	default:
 		return "libllama-embedder.so"
 	}

--- a/bindings/go/llama_embedder.go
+++ b/bindings/go/llama_embedder.go
@@ -3,9 +3,11 @@ package llama_embedder
 /*
 #cgo CFLAGS: -I.
 #cgo CXXFLAGS: -I. -std=c++11 -Wall -Wextra -pedantic
-#cgo !clang, !darwin LDFLAGS: -ldl -lstdc++
-#cgo darwin LDFLAGS: -ldl -stdlib=libc++
-#cgo clang LDFLAGS: -ldl -stdlib=libc++
+
+#cgo darwin LDFLAGS: -stdlib=libc++
+#cgo linux !clang LDFLAGS: -ldl -lstdc++
+#cgo linux clang LDFLAGS: -ldl -stdlib=libc++
+#cgo windows LDFLAGS: -lkernel32
 #include <stdlib.h>
 #include "wrapper.h"
 

--- a/bindings/go/llama_embedder_test.go
+++ b/bindings/go/llama_embedder_test.go
@@ -13,9 +13,8 @@ const defaultModelFile = "all-MiniLM-L6-v2.Q4_0.gguf"
 
 func TestLlamaEmbedder(t *testing.T) {
 	sharedLibPath := os.Getenv("SHARED_LIB_PATH")
-	fmt.Printf("sharedLibPath: %s\n", sharedLibPath)
 	if sharedLibPath == "" {
-		t.Errorf("SHARED_LIB_PATH is not set")
+		sharedLibPath = "../../build/"
 	}
 	sharedLibFile := filepath.Join(sharedLibPath, getOSSharedLibName())
 	err := downloadHFModel(defaultHFRepo, defaultModelFile, defaultModelFile, "")

--- a/bindings/go/llama_embedder_test.go
+++ b/bindings/go/llama_embedder_test.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"github.com/stretchr/testify/require"
 	"os"
+	"path/filepath"
 	"testing"
 )
 
@@ -12,10 +13,11 @@ const defaultModelFile = "all-MiniLM-L6-v2.Q4_0.gguf"
 
 func TestLlamaEmbedder(t *testing.T) {
 	sharedLibPath := os.Getenv("SHARED_LIB_PATH")
+	fmt.Printf("sharedLibPath: %s\n", sharedLibPath)
 	if sharedLibPath == "" {
-		sharedLibPath = "../../build/"
+		t.Errorf("SHARED_LIB_PATH is not set")
 	}
-	sharedLibFile := fmt.Sprintf("%s/%s", sharedLibPath, getOSSharedLibName())
+	sharedLibFile := filepath.Join(sharedLibPath, getOSSharedLibName())
 	err := downloadHFModel(defaultHFRepo, defaultModelFile, defaultModelFile, "")
 	require.NoError(t, err, "Failed to download model")
 	t.Cleanup(func() {

--- a/bindings/go/wrapper.cpp
+++ b/bindings/go/wrapper.cpp
@@ -8,10 +8,37 @@
 #include <cstring>
 #include <cstdlib>
 #include <iostream>
-#include <dlfcn.h>
 #include <stdexcept>
 #include "../../src/embedder.h"
 #include "wrapper.h"
+
+#if defined(_WIN32) || defined(_WIN64)
+
+// Helper function to get the last error message on Windows
+std::string GetLastErrorAsString() {
+    DWORD errorMessageID = ::GetLastError();
+    if(errorMessageID == 0) {
+        return std::string(); // No error message has been recorded
+    }
+
+    LPSTR messageBuffer = nullptr;
+    size_t size = FormatMessageA(
+        FORMAT_MESSAGE_ALLOCATE_BUFFER |
+        FORMAT_MESSAGE_FROM_SYSTEM |
+        FORMAT_MESSAGE_IGNORE_INSERTS,
+        NULL,
+        errorMessageID,
+        MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+        (LPSTR)&messageBuffer,
+        0,
+        NULL
+    );
+
+    std::string message(messageBuffer, size);
+    LocalFree(messageBuffer);
+    return message;
+}
+#endif
 
 typedef llama_embedder * (*init_embedder_local_func)(const char *, uint32_t);
 typedef void (*free_embedder_local_func)(llama_embedder *);
@@ -19,7 +46,7 @@ typedef FloatMatrix (*embed_c_local_func)(llama_embedder *, const char  ** , siz
 typedef int (*get_metadata_c_local_func)(llama_embedder *,MetadataPair**, size_t*);
 typedef void (*free_metadata_c_local_func)(MetadataPair*, size_t);
 
-void * libh = NULL;
+lib_handle libh = NULL;
 llama_embedder * embedder = NULL;
 init_embedder_local_func init_embedder_f = NULL;
 free_embedder_local_func free_embedder_f = NULL;
@@ -38,60 +65,114 @@ void set_last_error(const char* error_message) {
     last_error = error_message;
 }
 
-void * load_library(const char * shared_lib_path){
+lib_handle load_library(const char * shared_lib_path){
     try {
+#if defined(_WIN32) || defined(_WIN64)
+        libh = LoadLibraryA(shared_lib_path);
+        if (!libh) {
+            std::string error_message = "Failed to load shared library: " + GetLastErrorAsString());
+            throw std::runtime_error(error_message);
+        }
+#else
         libh = dlopen(shared_lib_path, RTLD_LAZY);
         if (!libh) {
             std::string error_message = "Failed to load shared library: " + std::string(dlerror());
             throw std::runtime_error(error_message);
         }
-        init_embedder_f = (llama_embedder *(*)(const char *, uint32_t)) dlsym(libh, "init_embedder");
+#endif
+
+#if defined(_WIN32) || defined(_WIN64)
+        init_embedder_f = (init_embedder_local_func) GetProcAddress(libh, "init_embedder");
+        if (!init_embedder_f) {
+            std::string error_message = "Failed to load init_embedder function: " + GetLastErrorAsString();
+            throw std::runtime_error(error_message);
+        }
+#else
+        init_embedder_f = (init_embedder_local_func) dlsym(libh, "init_embedder");
         if (!init_embedder_f) {
             std::string error_message = "Failed to load init_embedder function: " + std::string(dlerror());
             throw std::runtime_error(error_message);
         }
+#endif
 
-        free_embedder_f = (void (*)(llama_embedder *)) dlsym(libh, "free_embedder");
-
+#if defined(_WIN32) || defined(_WIN64)
+        free_embedder_f = (free_embedder_local_func) GetProcAddress(libh, "free_embedder");
+        if (!free_embedder_f) {
+            std::string error_message = "Failed to load free_embedder function: " + GetLastErrorAsString();
+            throw std::runtime_error(error_message);
+        }
+#else
+        free_embedder_f = (free_embedder_local_func) dlsym(libh, "free_embedder");
         if (!free_embedder_f) {
             std::string error_message = "Failed to load free_embedder function: " + std::string(dlerror());
             throw std::runtime_error(error_message);
         }
+#endif
 
+#if defined(_WIN32) || defined(_WIN64)
+        embed_f = (embed_c_local_func) GetProcAddress(libh, "embed_c");
+        if (!embed_f) {
+            std::string error_message = "Failed to load embed function: " + GetLastErrorAsString();
+            throw std::runtime_error(error_message);
+        }
+#else
         embed_f = (embed_c_local_func) dlsym(libh, "embed_c");
-
         if (!embed_f) {
             std::string error_message = "Failed to load embed function: " + std::string(dlerror());
             throw std::runtime_error(error_message);
         }
+#endif
 
+#if defined(_WIN32) || defined(_WIN64)
+        get_metadata_f = (get_metadata_c_local_func) GetProcAddress(libh, "get_metadata_c");
+        if (!get_metadata_f) {
+            std::string error_message = "Failed to load get_metadata function: " + GetLastErrorAsString();
+            throw std::runtime_error(error_message);
+        }
+#else
         get_metadata_f = (get_metadata_c_local_func) dlsym(libh, "get_metadata_c");
-
         if (!get_metadata_f) {
             std::string error_message = "Failed to load get_metadata function: " + std::string(dlerror());
             throw std::runtime_error(error_message);
         }
+#endif
 
+
+#if defined(_WIN32) || defined(_WIN64)
+        free_metadata_f = (free_metadata_c_local_func) GetProcAddress(libh, "free_metadata_c");
+        if (!free_metadata_f) {
+            std::string error_message = "Failed to load free_metadata function: " + GetLastErrorAsString();
+            throw std::runtime_error(error_message);
+        }
+#else
         free_metadata_f = (free_metadata_c_local_func) dlsym(libh, "free_metadata_c");
-
         if (!free_metadata_f) {
             std::string error_message = "Failed to load free_metadata function: " + std::string(dlerror());
             throw std::runtime_error(error_message);
         }
-
+#endif
 
         return libh;
     } catch (const std::exception &e) {
         std::string error_message = "Failed to load shared library: " + std::string(e.what());
         set_last_error(error_message.c_str());
         if (libh != NULL) {
-            dlclose(libh);
+
+#if defined(_WIN32) || defined(_WIN64)
+            if (!FreeLibrary(libh)){
+                fprintf(stderr, "Failed to free library %lu\n", GetLastError());
+            }
+#else
+            if(dlclose(libh) != 0){
+                fprintf(stderr, "Failed to close library %s\n", dlerror());
+            }
+#endif
         }
         return NULL;
     }
 }
 
-int init_llama_embedder(char * model_path, uint32_t pooling_type = 1 ) {
+int init_llama_embedder(char * model_path, uint32_t pooling_type ) {
     if (!libh) {
         set_last_error("Shared library not loaded, use load_library first.");
         return -1;

--- a/bindings/go/wrapper.cpp
+++ b/bindings/go/wrapper.cpp
@@ -26,12 +26,12 @@ std::string GetLastErrorAsString() {
         FORMAT_MESSAGE_ALLOCATE_BUFFER |
         FORMAT_MESSAGE_FROM_SYSTEM |
         FORMAT_MESSAGE_IGNORE_INSERTS,
-        NULL,
+        nullptr,
         errorMessageID,
         MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
         (LPSTR)&messageBuffer,
         0,
-        NULL
+        nullptr
     );
 
     std::string message(messageBuffer, size);
@@ -46,13 +46,13 @@ typedef FloatMatrix (*embed_c_local_func)(llama_embedder *, const char  ** , siz
 typedef int (*get_metadata_c_local_func)(llama_embedder *,MetadataPair**, size_t*);
 typedef void (*free_metadata_c_local_func)(MetadataPair*, size_t);
 
-lib_handle libh = NULL;
-llama_embedder * embedder = NULL;
-init_embedder_local_func init_embedder_f = NULL;
-free_embedder_local_func free_embedder_f = NULL;
-embed_c_local_func embed_f = NULL;
-get_metadata_c_local_func get_metadata_f = NULL;
-free_metadata_c_local_func free_metadata_f = NULL;
+lib_handle libh = nullptr;
+llama_embedder * embedder = nullptr;
+init_embedder_local_func init_embedder_f = nullptr;
+free_embedder_local_func free_embedder_f = nullptr;
+embed_c_local_func embed_f = nullptr;
+get_metadata_c_local_func get_metadata_f = nullptr;
+free_metadata_c_local_func free_metadata_f = nullptr;
 
 static std::string last_error;
 
@@ -156,7 +156,7 @@ lib_handle load_library(const char * shared_lib_path){
     } catch (const std::exception &e) {
         std::string error_message = "Failed to load shared library: " + std::string(e.what());
         set_last_error(error_message.c_str());
-        if (libh != NULL) {
+        if (libh != nullptr) {
 
 #if defined(_WIN32) || defined(_WIN64)
             if (!FreeLibrary(libh)){
@@ -168,7 +168,7 @@ lib_handle load_library(const char * shared_lib_path){
             }
 #endif
         }
-        return NULL;
+        return nullptr;
     }
 }
 
@@ -194,13 +194,21 @@ void free_llama_embedder() {
     if (embedder) {
         free_embedder_f(embedder);
     }
-    if (libh) {
-        dlclose(libh);
+    if (libh != nullptr) {
+#if defined(_WIN32) || defined(_WIN64)
+        if (!FreeLibrary(libh)){
+            fprintf(stderr, "Failed to free library %lu\n", GetLastError());
+        }
+#else
+        if(dlclose(libh) != 0){
+            fprintf(stderr, "Failed to close library %s\n", dlerror());
+        }
+#endif
     }
 }
 
 FloatMatrixW llama_embedder_embed(const char** texts, size_t text_count, int32_t norm) {
-    FloatMatrixW fm = {NULL, 0, 0};
+    FloatMatrixW fm = {nullptr, 0, 0};
     try {
         std::vector<std::vector<float>> output;
         auto f1 = embed_f(embedder, texts, text_count, norm);
@@ -215,25 +223,25 @@ FloatMatrixW llama_embedder_embed(const char** texts, size_t text_count, int32_t
 }
 
 char** llama_embedder_get_metadata(size_t* size) {
-    MetadataPair* metadata_array = NULL;
-    char** metadata = NULL;
+    MetadataPair* metadata_array = nullptr;
+    char** metadata = nullptr;
     *size = 0;
 
-    if (get_metadata_f(embedder, &metadata_array, size) != 0 || metadata_array == NULL) {
+    if (get_metadata_f(embedder, &metadata_array, size) != 0 || metadata_array == nullptr) {
         fprintf(stderr, "Failed to get metadata\n");
-        return NULL;
+        return nullptr;
     }
 
     metadata = (char**)malloc(*size * sizeof(char*));
-    if (metadata == NULL) {
+    if (metadata == nullptr) {
         fprintf(stderr, "Failed to allocate memory for metadata\n");
         free(metadata_array);
         *size = 0;
-        return NULL;
+        return nullptr;
     }
 
     for (size_t i = 0; i < *size; i++) {
-        if (metadata_array[i].key == NULL || metadata_array[i].value == NULL) {
+        if (metadata_array[i].key == nullptr || metadata_array[i].value == nullptr) {
             fprintf(stderr, "Null key or value at index %zu\n", i);
             continue;
         }
@@ -241,7 +249,7 @@ char** llama_embedder_get_metadata(size_t* size) {
         size_t key_len = strlen(metadata_array[i].key);
         size_t value_len = strlen(metadata_array[i].value);
         metadata[i] = (char*)malloc(key_len + value_len + 2); // +2 for '=' and null terminator
-        if (metadata[i] == NULL) {
+        if (metadata[i] == nullptr) {
             fprintf(stderr, "Failed to allocate memory for metadata[%zu]\n", i);
             continue;
         }

--- a/bindings/go/wrapper.cpp
+++ b/bindings/go/wrapper.cpp
@@ -70,7 +70,7 @@ lib_handle load_library(const char * shared_lib_path){
 #if defined(_WIN32) || defined(_WIN64)
         libh = LoadLibraryA(shared_lib_path);
         if (!libh) {
-            std::string error_message = "Failed to load shared library: " + GetLastErrorAsString());
+            std::string error_message = "Failed to load shared library: " + GetLastErrorAsString();
             throw std::runtime_error(error_message);
         }
 #else

--- a/bindings/go/wrapper.h
+++ b/bindings/go/wrapper.h
@@ -1,12 +1,27 @@
 //
 // Created by Trayan Azarov on 13.09.24.
 //
-#include <stdint.h>
-#include <stdbool.h>
-#include <dlfcn.h>
-#include <stdio.h>
 #ifndef WRAPPER_H
 #define WRAPPER_H
+#if defined(_WIN32) || defined(_WIN64)
+#if defined(BUILDING_DLL)
+        #define EXPORT_SYMBOL __declspec(dllexport)
+    #else
+        #define EXPORT_SYMBOL __declspec(dllimport)
+    #endif
+#else
+#define EXPORT_SYMBOL __attribute__((visibility("default")))
+#endif
+#include <stdint.h>
+#include <stdbool.h>
+#if defined(_WIN32) || defined(_WIN64)
+#include <windows.h>
+    typedef HMODULE lib_handle;
+#else
+#include <dlfcn.h>
+    typedef void* lib_handle;
+#endif
+#include <stdio.h>
 
 
 
@@ -20,16 +35,16 @@ typedef struct {
     size_t cols;
 } FloatMatrixW;
 
-void *load_library(const char *shared_lib_path);
-int init_llama_embedder(char *model_path, uint32_t pooling_type);
-void free_llama_embedder();
-FloatMatrixW llama_embedder_embed(const char **texts, size_t text_count, int32_t norm);
-void free_float_matrixw(FloatMatrixW * fm);
+EXPORT_SYMBOL lib_handle load_library(const char *shared_lib_path);
+EXPORT_SYMBOL int init_llama_embedder(char *model_path, uint32_t pooling_type);
+EXPORT_SYMBOL void free_llama_embedder();
+EXPORT_SYMBOL FloatMatrixW llama_embedder_embed(const char **texts, size_t text_count, int32_t norm);
+EXPORT_SYMBOL void free_float_matrixw(FloatMatrixW * fm);
 
-const char* get_last_error();
+EXPORT_SYMBOL const char* get_last_error();
 
-char ** llama_embedder_get_metadata(size_t* size);
-void free_metadata(char** metadata_array, size_t size);
+EXPORT_SYMBOL char ** llama_embedder_get_metadata(size_t* size);
+EXPORT_SYMBOL void free_metadata(char** metadata_array, size_t size);
 
 #ifdef __cplusplus
 }

--- a/bindings/go/wrapper.h
+++ b/bindings/go/wrapper.h
@@ -4,23 +4,16 @@
 #ifndef WRAPPER_H
 #define WRAPPER_H
 #if defined(_WIN32) || defined(_WIN64)
-#if defined(BUILDING_DLL)
-        #define EXPORT_SYMBOL __declspec(dllexport)
-    #else
-        #define EXPORT_SYMBOL __declspec(dllimport)
-    #endif
-#else
-#define EXPORT_SYMBOL __attribute__((visibility("default")))
-#endif
-#include <stdint.h>
-#include <stdbool.h>
-#if defined(_WIN32) || defined(_WIN64)
+#define EXPORT_SYMBOL __declspec(dllexport)
 #include <windows.h>
     typedef HMODULE lib_handle;
 #else
+#define EXPORT_SYMBOL __attribute__((visibility("default")))
 #include <dlfcn.h>
     typedef void* lib_handle;
 #endif
+#include <stdint.h>
+#include <stdbool.h>
 #include <stdio.h>
 
 

